### PR TITLE
[WIP] releng: Delete 1.16 branch jobs and shift generic version markers

### PR DESCRIPTION
**Pull request by justaugustus**
(Cleanup spawned from https://kubernetes.slack.com/archives/CJH2GBF7Y/p1604516450135900 and part of https://github.com/kubernetes/sig-release/issues/850.)

- releng: Remove 1.16 branch jobs

- releng: Shift version markers for generated jobs

  Generated jobs containing generic name / marker they now use:
  - beta / `latest` (master - 1.20)
  - stable1 / `latest-1.19`
  - stable2 / `latest-1.18`
  - stable3 / `latest-1.17`

- generated: Run `./hack/update-generated-tests.sh`

- releng: Remove 1.16 dashboards

- releng: Shift beta version marker creation to primary build job

- generated: Run config-rotator to shift release branch version markers

  ```console
  go run releng/config-rotator/main.go \
    --config-file config/jobs/kubernetes/sig-release/release-branch-jobs/1.19.yaml \
    --old beta \
    --new stable1
  go run releng/config-rotator/main.go \
    --config-file config/jobs/kubernetes/sig-release/release-branch-jobs/1.18.yaml \
    --old stable1 \
    --new stable2
  go run releng/config-rotator/main.go \
    --config-file config/jobs/kubernetes/sig-release/release-branch-jobs/1.17.yaml \
    --old stable2 \
    --new stable3
  ```

- releng: Allow build-canary job to be forked

  This is just a safeguard to ensure we don't disrupt any jobs targeting
  the K8s Infra version markers if the build job migration isn't complete
  ahead of the 1.20 branch cut.

### CODE REVIEW

**Comment by justaugustus** [comment]
I've opened two PRs to pull some commits out of this one:
- images(kubekins-e2e,krte): Stop building 1.16 images --> https://github.com/kubernetes/test-infra/pull/19853
- Delete WG Resource Management dashboard as it is now retired --> https://github.com/kubernetes/test-infra/pull/19854

**Comment by justaugustus** [comment]
Version markers are always pain...

With some of the markers shifted, some generated jobs have been elevated to release-blocking, which is now causing the following errors:

```console
--- FAIL: TestReleaseBlockingJobsMustHaveTestgridDescriptions (0.00s)
    config_test.go:533: dashboard_tab sig-release-master-blocking/gce-cos-k8sbeta-alphafeatures is release-blocking: - Must have alert_options (ensure blocking dashboard is listed first in testgrid-dashboards)
    config_test.go:533: dashboard_tab sig-release-master-blocking/gce-cos-k8sbeta-default is release-blocking: - Must have alert_options (ensure blocking dashboard is listed first in testgrid-dashboards)
    config_test.go:533: dashboard_tab sig-release-master-blocking/gce-cos-k8sbeta-ingress is release-blocking: - Must have alert_options (ensure blocking dashboard is listed first in testgrid-dashboards)
    config_test.go:533: dashboard_tab sig-release-master-blocking/gce-cos-k8sbeta-reboot is release-blocking: - Must have alert_options (ensure blocking dashboard is listed first in testgrid-dashboards)
```

I still need to think through fixing that a little, but the PR is ready for initial review from @kubernetes/release-engineering.

**Comment by justaugustus** [comment]
Ultimately, I think it may soon be time to revisit this PR, which removes the `beta` generated jobs: https://github.com/kubernetes/test-infra/pull/18169

**Comment by justaugustus** [comment]
/assign @spiffxp 
To get your thoughts on the generated jobs...

File: config/jobs/kubernetes/generated/generated.yaml [hunk 1]
```diff
@@ -1621,7 +1621,7 @@ periodics:
   cluster: k8s-infra-prow-build
   annotations:
     testgrid-tab-name: gce-cos-k8sbeta-reboot
-    testgrid-dashboards: sig-release-1.19-blocking
+    testgrid-dashboards: sig-release-master-blocking
```
Comment (spiffxp) [inline config/jobs/kubernetes/generated/generated.yaml:1624]: Why is a job named `beta` on master-blocking?

File: config/jobs/kubernetes/sig-release/release-branch-jobs/1.17.yaml [hunk 2]
```diff
@@ -1200,11 +1191,11 @@ presubmits:
         name: main
         resources:
           limits:
-            cpu: 2
-            memory: 1.2Gi
+            cpu: "2"
+            memory: 1288490188800m
```
Comment (spiffxp) [inline config/jobs/kubernetes/sig-release/release-branch-jobs/1.17.yaml:1195]: Can we leave this in the units it was in before?

File: releng/test_config.yaml [hunk 3]
```diff
@@ -522,21 +522,21 @@ k8sVersions:
     - --extract=ci/latest
     version: master
   beta:
+    args:
+    - --extract=ci/latest
+    version: master
```
Comment (spiffxp) [inline releng/test_config.yaml:527]: I don't think updating this makes sense until we have `ci/latest-1.20`

**Review by spiffxp — CHANGES_REQUESTED** [review]
My read is this is going to result in a bunch of `beta` jobs ending up on release-master-blocking, in addition to their `master` variants.  I would hold off on that.

I am also, fwiw, open to the idea of replacing the version marker names (both in job names and job parameters) with their concrete release-branch equivalents

**Comment by justaugustus** [comment]
Hah, that "needs rebase" popped up faster than I expected it to.

File: releng/test_config.yaml (hunk 3 above)
Comment (justaugustus) [reply releng/test_config.yaml:527]: @spiffxp -- Not quite. This is vestigial from when release branches were cut during the beta.0.

Now that we cut branches at rc.0 and cut betas from `master`, this is correct.
[v1.20.0-beta.1](https://github.com/kubernetes/kubernetes/releases/tag/v1.20.0-beta.1) was just released on Tuesday, so the "beta" jobs are essentially behind a version.

This was the motivation for introducing a `stable4` marker in https://github.com/kubernetes/test-infra/pull/18169:

> * Drop test config for beta jobs, add stable4 jobs
>   beta markers shift depending on the stage in the release cycle and can
>   end up being inaccurate without consistent care and feeding of the
>   generated jobs.
>   Here we remove the beta job configs (not the k8s-beta version marker;
>   that'll happen in an upcoming commit) and establish a new `stable4`
>   version config.
>   This serves a few purposes:
>   
>   * beta releases are now cut from the main (`master`) branch, so the
>     `dev` `k8sVersions`/`nodeK8sVersions` configs are now sufficient to
>     describe the beta phase of the release cycle
>   * now that we've expanded the support cycle for Kubernetes versions
>     (from 1.19 on) to 12+ months, removing the previous jobs no longer
>     falls neatly in line with the timing of the next branch cut. The
>     `stable4` job configs will now allow us to maintain generated jobs
>     through the newly extended support cycle.

File: config/jobs/kubernetes/sig-release/release-branch-jobs/1.17.yaml (hunk 2 above)
Comment (justaugustus) [reply config/jobs/kubernetes/sig-release/release-branch-jobs/1.17.yaml:1195]: Yeah, this was generated from `./hack/update-generated-tests.sh`.
We'd need to tweak that script to conform.

File: config/jobs/kubernetes/generated/generated.yaml (hunk 1 above)
Comment (justaugustus) [reply config/jobs/kubernetes/generated/generated.yaml:1624]: This is a result of alpha and beta releases being cut on `master` now :/
The generation script + `test_config.yaml` would need some refactor.

Comment (spiffxp) [reply config/jobs/kubernetes/generated/generated.yaml:1624]: I'm suggesting that refactor should be done as part of this PR, or we should wait until we have 1.20 dashboards to hold the beta jobs.  Adding both beta and master jobs to release-master-blocking isn't appropriate

**Comment by justaugustus** [comment]
I'll think through this after the break.
/close

**Pull request closed without merging**